### ISSUE TITLE
API for registering custom icons for the editor

### ISSUE THREAD

**Issue by getdave** [WordPress/gutenberg#51563]
There have been quite a few requests to allow us to customise the icons used in blocks in the editor

- https://github.com/WordPress/gutenberg/issues/51118
- https://github.com/WordPress/gutenberg/issues/37930
- https://github.com/WordPress/gutenberg/issues/16513

Currently this is not possible.

Whilst exploring this in https://github.com/WordPress/gutenberg/pull/51527 [I suggested](https://github.com/WordPress/gutenberg/pull/51527#issuecomment-1594271846) exploring a new API to allow developers to register their own custom icons with the editor.

```php
register_editor_icon(
    name: 'myicon',
    label: 'My Icon',
    svg: // SVG markup here
);
```

[As I suggested](https://github.com/WordPress/gutenberg/pull/51527#issuecomment-1594271846) we would need to register this PHP side and sanitize the SVG there.

Then on output we would:

- PHP: use `esc_` functions.
- JS: use `dangerouslySetInnerHTML` with equivalent escape functions.

**Comment by MaggieCabrera** [comment]
Related https://github.com/WordPress/gutenberg/issues/16513

**Comment by getdave** [comment]
@scruffian Is there any scope here to absorb icons as part of global styles as some kind of element? Is that even possible?

**Comment by MaggieCabrera** [comment]
I think it makes more sense for it to be [a block](https://github.com/WordPress/gutenberg/issues/16484)?

**Comment by getdave** [comment]
> I think it makes more sense for it to be [a block](https://github.com/WordPress/gutenberg/issues/16484)?

I agree that having a block for "icon" might be useful. However, this issue is specifically about providing a means for developers to use custom icons within the editor.

What we need is a means to register those icons (note: that doesn't mean we need to be able to upload them).

One use case for such an API can be found [here](https://github.com/WordPress/gutenberg/issues/37930).

**Comment by getdave** [comment]
[Prior art for Media Categories](https://github.com/WordPress/gutenberg/pull/51542/)

**Comment by scruffian** [comment]
> Is there any scope here to absorb icons as part of global styles as some kind of element? Is that even possible?

It's possible to create an icon element, but I don't think it would help much with what you're seeking to achieve here. It could be a good followup though.

**Comment by getdave** [comment]
I think we should consider doing this via Theme JSON instead. I'm going to think about this some more and come up with a counter proposal.

**Comment by MaggieCabrera** [comment]
@getdave did you ever get to explore the theme.json route? I wonder if we should do both ways?

**Comment by getdave** [comment]
No sorry and I won't have bandwidth to any time soon. Apologies

**Comment by youknowriad** [comment]
@t-hamano didn't you have a PR that address this? how is it going?

**Comment by t-hamano** [comment]
@youknowriad I've put it on hold because I had to work on 6.9, but here it is: https://github.com/WordPress/gutenberg/pull/72215

It's basically finished, but I may plan to update it soon. The initial release might be better developed under the experimental flag.
